src/backend/distributed/executor/local_executor.c

```diff
@@ -431,6 +407,39 @@ ExtractParametersForLocalExecution(ParamListInfo paramListInfo, Oid **parameterT
 }
 
 
+/*
+ * LocallyExecuteUtilityTask runs a utility command via local execution.
+ */
+static void
+LocallyExecuteUtilityTask(Task *task)
+{
+	/*
+	 * If we roll back to a savepoint, we may no longer be in a query on
+	 * a shard. Reset the value as we go back up the stack.
+	 */
+	uint64 prevInLocalQueryOnShard = InLocalQueryOnShard;
+
+	if (task->anchorShardId != INVALID_SHARD_ID)
+	{
+		InLocalQueryOnShard = task->anchorShardId;
```

Comment:
The `INSERT` command does not generally get here.

The `SELECT func(id) FROM dist` part will get broken down into tasks on shards, and the tasks will either be remote or local with InLocalQueryOnShard set to the distributed table shard ID, so func(id) would fail if it attempted to do a distributed query. We then separately copy into `ref`.

Something we don't currently capture is function calls that happen as part of a local copy operation (e.g. constraint has a function call that does a distributed query), though they would still fail if some of the rows are remote. 

Probably still useful to have some insert tests.

Reply:
added insert into .. select tests